lntest/harness.go

```diff
@@ -746,102 +740,42 @@ func saveProfilesPage(node *HarnessNode) error {
-// requests for the broadcast of a target transaction, and then dispatches the
-// transaction once its seen on the Bitcoin network.
-func (n *NetworkHarness) networkWatcher() {
-	seenTxns := make(map[chainhash.Hash]struct{})
-	clients := make(map[chainhash.Hash][]chan struct{})
-
-	for {
-
-		select {
-		case <-n.quit:
-			return
-
-		case req := <-n.bitcoinWatchRequests:
-			// If we've already seen this transaction, then
-			// immediately dispatch the request. Otherwise, append
-			// to the list of clients who are watching for the
-			// broadcast of this transaction.
-			if _, ok := seenTxns[req.txid]; ok {
-				close(req.eventChan)
-			} else {
-				clients[req.txid] = append(clients[req.txid], req.eventChan)
-			}
-		case txid := <-n.seenTxns:
-			// Add this txid to our set of "seen" transactions. So
-			// we're able to dispatch any notifications for this
-			// txid which arrive *after* it's seen within the
-			// network.
-			seenTxns[*txid] = struct{}{}
-
-			// If there isn't a registered notification for this
-			// transaction then ignore it.
-			txClients, ok := clients[*txid]
-			if !ok {
-				continue
-			}
-
-			// Otherwise, dispatch the notification to all clients,
-			// cleaning up the now un-needed state.
-			for _, client := range txClients {
-				close(client)
-			}
-			delete(clients, *txid)
-		}
-	}
-}
-
-// OnTxAccepted is a callback to be called each time a new transaction has been
-// broadcast on the network.
-func (n *NetworkHarness) OnTxAccepted(hash *chainhash.Hash) {
-	select {
-	case n.seenTxns <- hash:
-	case <-n.quit:
-		return
-	}
-}
-
-// WaitForTxBroadcast blocks until the target txid is seen on the network. If
+// WaitForTxInMempool blocks until the target txid is seen in the mempool. If
 // the transaction isn't seen within the network before the passed timeout,
 // then an error is returned.
-// TODO(roasbeef): add another method which creates queue of all seen transactions
-func (n *NetworkHarness) WaitForTxBroadcast(ctx context.Context, txid chainhash.Hash) error {
+func (n *NetworkHarness) WaitForTxInMempool(ctx context.Context,
+	txid chainhash.Hash) error {
+
 	// Return immediately if harness has been torn down.
 	select {
 	case <-n.quit:
 		return fmt.Errorf("NetworkHarness has been torn down")
 	default:
 	}
 
-	eventChan := make(chan struct{})
+	ticker := time.NewTicker(50 * time.Millisecond)
+	defer ticker.Stop()
 
-	n.bitcoinWatchRequests <- &txWatchRequest{
-		txid:      txid,
-		eventChan: eventChan,
-	}
+	var err error
+	var mempool []*chainhash.Hash
+	for {
+		select {
+		case <-ctx.Done():
+			return fmt.Errorf("wanted %v, found %v txs "+
+				"in mempool: %v", txid, len(mempool), mempool)
 
-	select {
-	case <-eventChan:
-		return nil
-	case <-n.quit:
-		return fmt.Errorf("NetworkHarness has been torn down")
-	case <-ctx.Done():
-		return fmt.Errorf("tx not seen before context timeout")
+		case <-ticker.C:
+			mempool, err = n.Miner.Node.GetRawMempool()
```

Comment:
Isn't it possible we never find the tx in the mempool if "mining" happens on another goroutine? I guess it's not a problem right now, but it could be?

Reply:
I'm not sure I understand what you mean. This is asking the miner node directly to get its mempool. If the miner node doesn't have a TX, it cannot get mined (only that node is creating blocks). Or do you mean that it is mined before we even start looking for it? Then the test is written incorrectly and will fail (mining only ever happens manually in the itests).

Reply:
What I meant is that between two snapshots of the mempool the miner could theoretically mine the tx, but if everything is sequential this is a non issue (in the tests).